scripts/3_hw4_solution_multiple_cores.r

```diff
@@ -0,0 +1,142 @@
+# Assignment 1:
+library(tweedie)
+library(ggplot2)
+
+simTweedieTest <-  
+  function(N){ 
+    t.test( 
+      rtweedie(N, mu=10000, phi=100, power=1.9), 
+      mu=10000 
+    )$p.value 
+  } 
+
+# Assignment 2:
+MTweedieTests <-  
+  function(N, M, sig){ 
+    library(parallel)  # Load the parallel package
+    # Split M simulations into multiple cores
+    num_cores <- detectCores()
+    M_per_core <- ceiling(M / num_cores)
+    results <- mclapply(1:num_cores, function(i) {
+      replicate(M_per_core, simTweedieTest(N))
+    })
+    
+    # Combine the results from all cores
+    all_results <- unlist(results)
+    sum(all_results < sig) / M
+  } 
```

Comment:
I am not familiar with the mclapply() function for parallization, but for me this one took 6 seconds less than the second alternative. I guess it is correct then:) 

Here is a similar solution using the foreach() function:
```r
MParTweedieTests <-
  function(N, M, sig, maxcores) {
    Cores <- min(parallel::detectCores(), maxcores)
    
    cl <- makeCluster(Cores)
    registerDoParallel(cl)
    
    m_per_core <- round(M / Cores)
    m_total <- m_per_core * Cores
    
    rejects_per_core <-
      foreach(
        i = 1:Cores,
        .combine = 'rbind',
        .packages = c('magrittr', 'dplyr', 'tweedie'),
        .export = c('simTweedieTest')
      ) %dopar%
      sum(replicate(m_per_core, simTweedieTest(N)) < sig)
    
    stopCluster(cl)
    
    sum(rejects_per_core) / m_total
  }
```